DQMServices/FwkIO/plugins/DQMRootSource.cc

```diff
@@ -80,21 +80,26 @@ namespace {
         edm::LogError("MergeFailure")<<"Failed to merge DQM element "<<iOriginal->GetName();
       }
     } else {
-      if (iOriginal->GetNbinsX() == iToAdd->GetNbinsX() &&
-          iOriginal->GetXaxis()->GetXmin() == iToAdd->GetXaxis()->GetXmin() &&
-          iOriginal->GetXaxis()->GetXmax() == iToAdd->GetXaxis()->GetXmax() &&
-          iOriginal->GetNbinsY() == iToAdd->GetNbinsY() &&
-          iOriginal->GetYaxis()->GetXmin() == iToAdd->GetYaxis()->GetXmin() &&
-          iOriginal->GetYaxis()->GetXmax() == iToAdd->GetYaxis()->GetXmax() &&
-          iOriginal->GetNbinsZ() == iToAdd->GetNbinsZ() &&
-          iOriginal->GetZaxis()->GetXmin() == iToAdd->GetZaxis()->GetXmin() &&
-          iOriginal->GetZaxis()->GetXmax() == iToAdd->GetZaxis()->GetXmax() &&
-	  MonitorElement::CheckBinLabels(iOriginal->GetXaxis(),iToAdd->GetXaxis()) &&
-	  MonitorElement::CheckBinLabels(iOriginal->GetYaxis(),iToAdd->GetYaxis()) &&
-	  MonitorElement::CheckBinLabels(iOriginal->GetZaxis(),iToAdd->GetZaxis())) {
-	iOriginal->Add(iToAdd);
-      } else {
-	edm::LogError("MergeFailure")<<"Found histograms with different axis limits or different labels'"<<iOriginal->GetName()<<"' not merged.";
+      try {
+        // TODO: we might want to switch to Merge() which should be smarter. It
+        // seems that Add() would call Merge() if needed (not identical bins
+        // etc.) but that seems to not work, see below.
+        iOriginal->Add(iToAdd);
```

Comment:
Do we have any idea if TH1::Add is exception safe? That is, after an exception is thrown, is the TH1 left in a valid state?

Reply:
That is a very good point. I'd think so, but we should check what ROOT has to say. Actually, I am no longer fully certain that `Add` is supposed to throw at all...

We should certainly check the return value now. One moment please...

Reply:
ROOT doesn't cause the throw, CMS does. By default, we convert ROOT error messages into exceptions (by plugging into ROOT's error handling system). We've done this for more than a decade as it has been the only reasonable way to quickly spot problems originating in ROOT.

Reply:
Ah, that explains the mysterious step that was not clear to me. Reading the error message actually points to a specific place in `TH1Merger`, let me check what is going on there.